url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/ui/timeaxis/TimeAxisUtils.java

```diff
@@ -0,0 +1,87 @@
+package org.swiftgantt.ui.timeaxis;
+
+
+import org.swiftgantt.Config;
+import org.swiftgantt.GanttChart;
+import org.swiftgantt.common.CalendarUtils;
+import org.swiftgantt.common.Time;
+import org.swiftgantt.model.Task;
+import org.swiftgantt.ui.TimeUnit;
+
+/**
+ * Utils for time axis.
+ * 
+ * @author Yuxing Wang
+ * @version 1.0
+ */
+public class TimeAxisUtils {
+
+	/**
+	 * Calculate interval between actual start time and actual end time of a task.
+	 * 
+	 * @param tu
+	 * @param task
+	 * @return
+	 */
+	public static int getActualTimeIntervalByTimeUnit(TimeUnit tu, Task task) {
+		return getTimeIntervalByTimeUnit(tu, task.getActualStart(), task.getActualEnd());
+	}
+
+	/**
+	 * Including rest-out time.
+	 * 
+	 * @param timeUnit
+	 * @param start
+	 * @param end
+	 * @return
+	 */
+	public static int getTimeIntervalByTimeUnit(TimeUnit timeUnit, Time start, Time end) {
+		int ret = 0;
+		if (timeUnit == TimeUnit.Hour) {
+			// Serveral rest-out hours merge to one step.
+			Config cfg = GanttChart.getStaticConfig();
+			int startHour = cfg.getWorkingHoursSpanOfDay()[0];
+			int hours = cfg.getWorkingHoursOfDay();
+			ret = CalendarUtils.calcWorkingHourDurationLen(start, end, hours, startHour) - 1;// -1 because need return interval, not duration
+		} else if (timeUnit == TimeUnit.AllDay) {
+			ret = start.getHourIntervalTo(end);//CalendarUtils.getHourInterval(start, end);
+		} else if (timeUnit == TimeUnit.Day) {
+			ret = start.getDayIntervalTo(end);//CalendarUtils.getDateInterval(start, end);
+		} else if (timeUnit == TimeUnit.Week) {
+			ret = start.getWeekIntervalTo(end);//CalendarUtils.getWeekInterval(start, end);
+		} else if (timeUnit == TimeUnit.Month) {
+			ret = start.getMonthIntervalTo(end);//CalendarUtils.getMonthInterval(start, end);
+		} else if (timeUnit == TimeUnit.Year) {
+			ret = start.getYearIntervalTo(end);//CalendarUtils.getYearInterval(start, end);
+		}
+		return ret;
+	}
+
+	/**
+	 * Calculate accurate time portion by time unit for accurate task bar rendering.
+	 * 
+	 * @since 0.3.2
+	 * @param timeUnit
+	 * @param time
+	 * @return
+	 */
+	public static float calcAccurateTimePortionByTimeUnit(TimeUnit timeUnit, Time time) {
+		float ret = 0;
+		if (timeUnit == TimeUnit.Hour) {
+			ret = (time.get(Time.MINUTE) + 1) / 60F;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;time&#39; (index: 1 | type: Time). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;timeUnit&#39;. The same parameter type is validated here for example: url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/GanttChart.java:359. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).